### PR TITLE
[parallel execution] delta support

### PR DESCRIPTION
**Pull request by georgemitenkov**
Introduces Delta handling inside parallel execution.
* MVHashMap now contains deltas as well as full writes. Reads proactively iterate downwards until any consecutive deltas are resolved (if no write is found, the delta is resolved with a base value from storage).  
* Validation covers aggregators, it ensures that the captured read value from MVHashMap remains the same, including a speculative error. Note that over-/underflows currently aren't supported, so if there is a real error, panic ensues when the outputs are prepared, i.e. when resolving the finalized deltas. Speculative error provides value 0 to the Move-VM, so the Move-VM can produce some output.
* Logic to treat consider delta writes inside the executor for write-outside & other considerations. Write outside is now updates outside.
* Preparing outputs happens sequentially, and moreover, with another linear traversal. This needs to be parallelized and combined with the output extraction (from last_input_outputs) in order to support more aggregators in the system.
* Adds support for delta testing in the non-Move-VM testing framework, new unit and prop tests passing.
* Adds OutputDeltaResolver, which should efficiently resolves deltas in parallel execution output, is testable and prop-tested in the fake PE testing framework, and should be parallelizable when there are more aggregators.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3126)

### CODE REVIEW

File: aptos-move/parallel-executor/src/executor.rs
Comment (sasha8) [inline aptos-move/parallel-executor/src/executor.rs:98]: We have an ABA here. I am convinced it is OK, but it is a bit scary.

File: aptos-move/parallel-executor/src/proptest_types/types.rs
Comment (sasha8) [inline aptos-move/parallel-executor/src/proptest_types/types.rs:325]: Is this code implemented in another diff?

**Review by sasha8 — COMMENTED** [review]
Looks good. Waiting for the missing code for approval.

File: aptos-move/parallel-executor/src/executor.rs
Comment (georgemitenkov) [reply aptos-move/parallel-executor/src/executor.rs:98]: I think this is not much different from resolved case?

File: aptos-move/mvhashmap/src/lib.rs
Comment (georgemitenkov) [reply aptos-move/mvhashmap/src/lib.rs:186]: Added a comment to Version enum definition

File: aptos-move/mvhashmap/src/unit_tests/mod.rs
Comment (gelash) [inline aptos-move/mvhashmap/src/unit_tests/mod.rs:29]: Can we add a few reads that test new cases (i.e. delta to storage, delta resolved, delta dependency, failure when we add my changes)?
Ideally to proptest as well but at least to the real unittests.

File: aptos-move/parallel-executor/src/executor.rs
Comment (sasha8) [reply aptos-move/parallel-executor/src/executor.rs:99]: Yes, both scary...

File: aptos-move/parallel-executor/src/executor.rs
Comment (sasha8) [inline aptos-move/parallel-executor/src/executor.rs:273]: What is this? We should take it from Rati's diff, right?

**Review by sasha8 — COMMENTED** [review]
Let's cherry-pick Rati's changes, test, and land it.

File: aptos-move/aptos-aggregator/src/delta_change_set.rs
Comment (zekun000) [inline aptos-move/aptos-aggregator/src/delta_change_set.rs:198]: #[cfg(any(test, feature = "fuzzing"))] is how we do cross-crate gating

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (zekun000) [inline aptos-move/aptos-vm/src/parallel_executor/mod.rs:104]: do we need this? I remember data_cache is per block

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:131]: map_or(true, ...)

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:218]: can we unify the name either way? accumulator or delta is fine

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:191]: it's more idiomatic to write like this
```
  match (aggregator.as_ref(), &entry.cell) => {
    (Some(delta), EntryCell::Write(..) => ..,
    (None, EntryCell::Write(..) => ..,
    (Some(delta), EntryCell::Delta(..) => ..,
    ...
  }
```

File: aptos-move/parallel-executor/src/executor.rs
Comment (zekun000) [inline aptos-move/parallel-executor/src/executor.rs:280]: this should always be false? or it's optimization for the case that actual value is coincidentally 0?

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs:104]: At the end of the parallel execution, currently, we still have output as a combination of writes and deltas, and need to transform the deltas into full writes. If we did it naively, resolving each delta could be a linear traversal (all the way to storage value or latest write), so instead we go bottom up and resolve it using data_cache, much like in sequential execution (but only for deltas). In other words, data cache must contains the resolved latest value for each path that previously had a delta.

There is an issue in this logic though, because we also have to make sure to push writes to the same paths that had a delta previously (currently it only pushes deltas). We test this case for PE but missed this stupid interconnect. I will fix it tonight.

File: aptos-move/parallel-executor/src/executor.rs
Comment (sasha8) [reply aptos-move/parallel-executor/src/executor.rs:280]: It is for the case the assumption does not hold and we have an overflow?

Comment (gelash) [reply aptos-move/parallel-executor/src/executor.rs:280]: what @sasha8 said (so we don't loop forever and instead panic later, when extracting outputs). Thinking is that panic is preferable to hanging, and it's also more easily testable (i.e. fake PE tests also test this case currently, that when there is an actual over/underflow, PE outputs materialized will also lead to overflow/underflow and we can panic there).

**Review by sasha8 — APPROVED** [review]
But how can you execute sequential workload in parallel??? Is this magic?

File: aptos-move/parallel-executor/src/executor.rs
Comment (zekun000) [reply aptos-move/parallel-executor/src/executor.rs:280]: please add a comment

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:202]: I think this can be a WriteOp::Delete and we have to handle it instead of panic

**Review by zekun000 — COMMENTED** [review]
also delete the proptest files

**Comment by gelash** [comment]
@zekun000 what do you mean delete proptest files?

**Comment by zekun000** [comment]
aptos-move/mvhashmap/proptest-regressions/unit_tests/proptest_types.txt

File: aptos-move/mvhashmap/src/lib.rs
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:202]: Fixed, and adjusted both PE and MVHashMap prop and unit tests. I think I verified, but @georgemitenkov should also double check that our "None" handling on aggregator read side results in an error (is the same case as read went to storage and there was nothing there).

File: aptos-move/mvhashmap/src/lib.rs
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs:191]: the logic changed somewhat where some conditions are shared and this might be a bit verbose.

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs:104]: I will unresolve till I fix the issue.

File: aptos-move/aptos-aggregator/src/delta_change_set.rs
Comment (gelash) [reply aptos-move/aptos-aggregator/src/delta_change_set.rs:198]: can't get it to work :/

File: aptos-move/mvhashmap/src/lib.rs
Comment (gelash) [reply aptos-move/mvhashmap/src/lib.rs]: done

File: aptos-move/parallel-executor/src/output_delta_resolver.rs
Comment (sasha8) [inline aptos-move/parallel-executor/src/output_delta_resolver.rs:49]: If this is a delete, then we will later panic. We need to document that we currently assume no deletion and deal with this later before enabling other use cases.

After discussion with Zekun, we are probably fine because we will have a read-write conflict accessing the aggregator. Need to test.

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:119]: the name is confusing

File: aptos-move/mvhashmap/src/lib.rs
Comment (zekun000) [inline aptos-move/mvhashmap/src/lib.rs:212]: this is what the vm would do https://github.com/aptos-labs/aptos-core/blob/634e76e8da009f7f7504764c312a94c8a047d378/aptos-move/aptos-aggregator/src/aggregator_extension.rs#L225

validation should correct it to abort on reading the aggregator resource (not the aggregator value).

**Review by sasha8 — APPROVED** [review]

**Review by zekun000 — APPROVED** [review]

File: aptos-move/parallel-executor/src/output_delta_resolver.rs
Comment (georgemitenkov) [reply aptos-move/parallel-executor/src/output_delta_resolver.rs:49]: I was discussing with @gelash : deletion is a tricky case since in Move we would consume a value, which means that the resource that holds aggregator (which we delete) must be deleted as well, right? Examples:
```
tx2: remove aggregator from table and remove it
tx3: borrow aggregator from table and add // fails on borrow!
```
I think we have the same thing with vector and option (implemented as vector under the hood)

Comment (georgemitenkov) [reply aptos-move/parallel-executor/src/output_delta_resolver.rs:49]: Or better say: whatever holds aggregator controls access to it. if we remove aggregator, we fail on accessing it via that holding resource, or fail when resource is also destroyed

File: aptos-move/mvhashmap/src/lib.rs
Comment (georgemitenkov) [reply aptos-move/mvhashmap/src/lib.rs:119]: +1

File: aptos-move/parallel-executor/src/output_delta_resolver.rs
Comment (zekun000) [reply aptos-move/parallel-executor/src/output_delta_resolver.rs:49]: yeah I think the deletion and delta op will be detected as write/read conflict on the container of the resource itself, but we can't panic because of the speculative execution

**Pull request merged**
